tests/callbacks/test_early_stopping.py

```diff
@@ -372,53 +377,122 @@ def test_early_stopping_mode_options():
 
 class EarlyStoppingModel(BoringModel):
 
-    def __init__(self, expected_end_epoch):
+    def __init__(self, expected_end_epoch: int, early_stop_on_train: bool):
         super().__init__()
         self.expected_end_epoch = expected_end_epoch
+        self.early_stop_on_train = early_stop_on_train
 
-    def validation_epoch_end(self, outputs):
+    def _epoch_end(self) -> None:
         losses = [8, 4, 2, 3, 4, 5, 8, 10]
-        val_loss = losses[self.current_epoch]
-        self.log('abc', torch.tensor(val_loss))
+        loss = losses[self.current_epoch]
+        self.log('abc', torch.tensor(loss))
         self.log('cba', torch.tensor(0))
 
+    def training_epoch_end(self, outputs):
+        if not self.early_stop_on_train:
+            return
+        self._epoch_end()
+
+    def validation_epoch_end(self, outputs):
+        if self.early_stop_on_train:
+            return
+        self._epoch_end()
+
     def on_train_end(self) -> None:
         assert self.trainer.current_epoch == self.expected_end_epoch, 'Early Stopping Failed'
 
 
 @pytest.mark.parametrize(
-    "callbacks, expected_stop_epoch, accelerator, num_processes",
+    "callbacks, expected_stop_epoch, check_on_train_epoch_end, accelerator, num_processes",
     [
-        ([EarlyStopping(monitor='abc'), EarlyStopping(monitor='cba', patience=3)], 3, None, 1),
-        ([EarlyStopping(monitor='cba', patience=3),
-          EarlyStopping(monitor='abc')], 3, None, 1),
-        pytest.param([EarlyStopping(monitor='abc'),
-                      EarlyStopping(monitor='cba', patience=3)],
-                     3,
-                     'ddp_cpu',
-                     2,
-                     marks=RunIf(skip_windows=True)),
-        pytest.param([EarlyStopping(monitor='cba', patience=3),
-                      EarlyStopping(monitor='abc')],
-                     3,
-                     'ddp_cpu',
-                     2,
-                     marks=RunIf(skip_windows=True)),
+        ([EarlyStopping(monitor='abc'), EarlyStopping(monitor='cba', patience=3)], 3, False, None, 1),
+        (
+            [EarlyStopping(monitor='cba', patience=3),
+             EarlyStopping(monitor='abc')],
+            3,
+            False,
+            None,
+            1,
+        ),
+        pytest.param(
+            [EarlyStopping(monitor='abc'), EarlyStopping(monitor='cba', patience=3)],
+            3,
+            False,
+            'ddp_cpu',
+            2,
+            marks=RunIf(skip_windows=True),
```

Comment:
why do you skip Windows? the CPU version of DDP shall be supported, right? @awaelchli